File: src/main/java/de/taschi/bulkgpxviewer/files/LoadedFileManager.java
package de.taschi.bulkgpxviewer.files;

/*-
 * #%L
 * bulkgpxviewer
 * %%
 * Copyright (C) 2021 S. Hillebrand
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import javax.swing.SwingUtilities;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import de.taschi.bulkgpxviewer.geo.GpxFile;
import io.jenetics.jpx.GPX;

/**
 * Singleton class for access to the loaded files and GPX tracks
 */
public class LoadedFileManager {
	
	private static final Logger LOG = LogManager.getLogger(LoadedFileManager.class);

	private static LoadedFileManager INSTANCE;

	private final List<GpxFile> loadedTracks = new ArrayList<>();
	
	private final List<LoadedFileChangeListener> changeListeners = new LinkedList<>();
	
	private GpxFile highlightedTrack = null;

	private Path loadedDirectory;

	public static LoadedFileManager getInstance() {
		if (INSTANCE == null) {
			INSTANCE = new LoadedFileManager();
		}
		return INSTANCE;
	}

	/**
	 * Load one GPX file
	 * @param file
	 */
	private void loadFile(Path file) {
		LOG.info("Loading GPX file " + file.toString()); //$NON-NLS-1$
		
		try {
			var gpx = GPX.read(file);		
			var track = new GpxFile(file, gpx);
			loadedTracks.add(track);
			
			LOG.info("GPX file " + file.toString() + " has been successfully loaded."); //$NON-NLS-1$ //$NON-NLS-2$
			
			fireChangeListeners();
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	public void fireChangeListeners() {
		SwingUtilities.invokeLater(() -> {
			for (LoadedFileChangeListener listener: changeListeners) {
				listener.onLoadedFileChange();
			}
		});
	}

	/**
	 * Remove all currently loaded GPX files and load all files from the given directory
	 * @param directory
	 */
	public synchronized void clearAndLoadAllFromDirectory(Path directory) throws IOException {		
		PathMatcher matcher = FileSystems.getDefault().getPathMatcher("glob:**.gpx"); //$NON-NLS-1$

		loadedTracks.clear();
		
		try {
			Files.walk(directory)
				.filter((it) -> matcher.matches(it))
				.forEach((it) -> loadFile(it));

			loadedDirectory = directory;
			new Thread(() -> fireChangeListeners()).start();
		} catch (RuntimeException e) {
			if (e.getCause() != null && e.getCause() instanceof IOException) {
				throw (IOException) e.getCause();
			} else {
				throw e;
			}
		}
	}
	
	/**
	 * Add a file change listener
	 * @param listener
	 */
	public void addChangeListener(LoadedFileChangeListener listener) {
		changeListeners.add(listener);
	}
	
	/**
	 * Remove a file change listener
	 * @param listener
	 */
	public void removeChangeListener(LoadedFileChangeListener listener) {
		changeListeners.remove(listener);
	}

	/**
	 * get all currently loaded tracks
	 * @return
	 */
	public synchronized List<GpxFile> getLoadedTracks() {
		return Collections.unmodifiableList(loadedTracks);
	}
	
	public synchronized void setHighlightedTrack(GpxFile track) {
		highlightedTrack = track;
	}
	
	public synchronized GpxFile getHighlightedTrack() {
		return highlightedTrack;
	}

	/**
	 * Reloads files from current directory
	 */
	public void refresh() throws IOException {
		clearAndLoadAllFromDirectory(loadedDirectory);
	}
}

File: src/main/java/de/taschi/bulkgpxviewer/files/LoadedFileManager.kt
package de.taschi.bulkgpxviewer.files

import com.google.inject.Singleton
import de.taschi.bulkgpxviewer.files.LoadedFileManager
import io.jenetics.jpx.GPX
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.PathMatcher
import java.util.*
import java.util.function.Consumer
import java.util.function.Predicate
import javax.swing.SwingUtilities

/*-
 * #%L
 * bulkgpxviewer
 * %%
 * Copyright (C) 2021 S. Hillebrand
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

/**
 * Singleton class for access to the loaded files and GPX tracks
 */
@Singleton
class LoadedFileManager constructor() {
    private val loadedTracks: MutableList<GpxFile> = ArrayList()
    private val changeListeners: MutableList<LoadedFileChangeListener> = LinkedList()

    @get:Synchronized
    @set:Synchronized
    var highlightedTrack: GpxFile? = null
    private var loadedDirectory: Path? = null

    /**
     * Load one GPX file
     * @param file
     */
    private fun loadFile(file: Path) {
        log.info("Loading GPX file $file") //$NON-NLS-1$
        try {
            val gpx = GPX.read(file)
            val track = GpxFile(file, gpx)
            loadedTracks.add(track)
            log.info("GPX file $file has been successfully loaded.") //$NON-NLS-1$ //$NON-NLS-2$
            fireChangeListeners()
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun fireChangeListeners() {
        SwingUtilities.invokeLater(Runnable({
            for (listener: LoadedFileChangeListener in changeListeners) {
                listener.onLoadedFileChange()
            }
        }))
    }

    /**
     * Remove all currently loaded GPX files and load all files from the given directory
     * @param directory
     */
    @Synchronized
    @Throws(IOException::class)
    fun clearAndLoadAllFromDirectory(directory: Path?) {
        val matcher: PathMatcher = FileSystems.getDefault().getPathMatcher("glob:**.gpx") //$NON-NLS-1$
        loadedTracks.clear()
        try {
            Files.walk(directory)
                .filter(Predicate({ it: Path? -> matcher.matches(it) }))
                .forEach(Consumer({ it: Path -> loadFile(it) }))
            loadedDirectory = directory
            Thread(Runnable({ fireChangeListeners() })).start()
        } catch (e: RuntimeException) {
            if (e.cause != null && e.cause is IOException) {
                throw (e.cause as IOException?)!!
            } else {
                throw e
            }
        }
    }

    /**
     * Add a file change listener
     * @param listener
     */
    fun addChangeListener(listener: LoadedFileChangeListener) {
        changeListeners.add(listener)
    }

    /**
     * Remove a file change listener
     * @param listener
     */
    fun removeChangeListener(listener: LoadedFileChangeListener) {
        changeListeners.remove(listener)
    }

    /**
     * get all currently loaded tracks
     * @return
     */
    @Synchronized
    fun getLoadedTracks(): List<GpxFile> {
        return Collections.unmodifiableList(loadedTracks)
    }

    /**
     * Reloads files from current directory
     */
    @Throws(IOException::class)
    fun refresh() {
        clearAndLoadAllFromDirectory(loadedDirectory)
    }

    companion object {
        private val log = LoggerFactory.getLogger(LoadedFileManager::class.java)
    }
}